transit/script/muni.py

```diff
@@ -0,0 +1,148 @@
+        special_names = {
+            "J": "J-Church ",
+            "K": "KT-Ingleside/Third Street ",
+            "L": "L-Taraval ",
+            "M": "M-Ocean View ",
+            "N": "N-Judah ",
+        }
+        return special_names.get(line, line)
+    else:
+        return line
+
+
+def map_name_to_direction(name):
+    if name.endswith("I"):
+        return "IB"
+    elif name.endswith("O"):
+        return "OB"
+    else:
+        return None  # Return None for other cases
+
+
+def process_muni(
+    file_name,
+    model_run_dir,
+    transit_line_rename_filepath,
+    transit_input_dir,
+    observed_MUNI_Line,
+    output_transit_dir,
+    model_MUNI_Line,
+):
+    line_names = read_transit_lines(model_run_dir, transit_line_rename_filepath)
+
+    MUNI = read_dbf_and_groupby_sum(
+        file_name, "SF MUNI", ["FULLNAME", "NAME", "TOD"], "AB_BRDA"
+    )
+    MUNI["Direction"] = MUNI["NAME"].apply(map_name_to_direction)
+
+    MUNI = MUNI.sort_values(by="FULLNAME").reset_index(drop=True)
+    MUNI = MUNI.rename(columns={"NAME": "Name", "AB_BRDA": "Ridership"})
+
+    MUNI_full = pd.merge(MUNI, line_names, on="Name", how="left")
+
+    # Apply the transformation function to the 'Line' column
+    MUNI_full["Line"] = MUNI_full["Line"].apply(transform_line)
+    obs_MUNI_line = pd.read_csv(transit_input_dir / observed_MUNI_Line)
+    mode = obs_MUNI_line[["Line", "Mode"]].drop_duplicates().reset_index(drop=True)
+    mode_dict = mode.set_index("Line")["Mode"].to_dict()
+    MUNI_full["Mode"] = MUNI_full["Line"].map(mode_dict)
+    MUNI_full["Key_line_dir"] = MUNI_full["Line"].astype(str) + MUNI_full["Direction"]
+    MUNI_full["Key_line_tod"] = (
+        MUNI_full["Line"].astype(str) + MUNI_full["TOD"] + MUNI_full["Direction"]
+    )
+    MUNI_full = MUNI_full[
+        [
+            "Line",
+            "Mode",
+            "Direction",
+            "TOD",
+            "Key_line_dir",
+            "Key_line_tod",
+            "Ridership",
+            "Name",
+        ]
+    ]
+    MUNI_full = MUNI_full.sort_values(by=["Line", "Direction", "TOD"]).reset_index(
+        drop=True
+    )
+    MUNI_full.to_csv(output_transit_dir / model_MUNI_Line, index=False)
+
+
+if __name__ == "__main__":
+    with open("transit.toml", "rb") as f:
+        config = tomllib.load(f)
+
+    transit_line_rename_filepath = (
+        Path(config["directories"]["resources"])
+        / config["transit"]["line_rename_filename"]
+    )
+    model_run_dir = config["directories"]["model_run"]
+    model_MUNI_Line = config["output"]["model_MUNI_Line"]
+    observed_MUNI_Line = Path(config["transit"]["observed_MUNI_Line"])
+    transit_input_dir = Path(config["directories"]["transit_input_dir"])
+    output_dir = model_run_dir / "validation_workbook" / "output"
+    output_transit_dir = output_dir / "transit"
+    output_transit_dir.mkdir(parents=True, exist_ok=True)
+    time_periods = ["EA", "AM", "MD", "PM", "EV"]
```

Comment:
all the time periods maybe should be imported from the utils script if they keep showing up in each of these py files

Reply:
@chunhochow I agree.